contents/handbook/engineering/query-performance-optimization.md

```diff
@@ -0,0 +1,49 @@
+---
+title: ClickHouse query performance optimization
+sidebar: Handbook
+showTitle: true
+---
+
+Making sure PostHog operates fast at scale is key to our success.
+
+This document outlines some tools we have to discover and fix issues.
+
+# Tools of the trade
+
+## 1. Grafana "Clickhouse queries - by endpoint" dashboard
+
+https://metrics.posthog.com/d/vo7oCVZ7z/clickhouse-queries-by-endpoint
+
+This dashboard gives a breakdown of how things are looking reliability and performance-wise.
+
+Highly used and slow/unreliable endpoints often indicate issues with queries.
+
+## 2. Instance status dashboard
+
+Under https://app.posthog.com/instance/status/internal_metrics you will find various metrics and query logs.
+
+Note that if you are a staff user you can also analyze queries by clicking on them (or copying your own queries).
+
+This analysis will output:
+- Query runtime
+- Number of rows read / Bytes read
+- Memory used
+- Flamegraphs for CPU, time and memory
+
+These can be useful for figuring out _why_ certain queries are performing slow.
```

Comment:
Are there any external resources we can link to on performance characteristics of clickhouse queries

Reply:
None that I'm aware of but we can build a list over time.